vale/code/crypto/sha/Vale.SHA.SHA_helpers.fst

```diff
@@ -174,12 +180,22 @@ let lemma_add_mod32_associates (x y z:int) :
   ()
 *)
 
-#reset-options "--using_facts_from ''"
 let lemma_add_mod_associates_U32 (x y z:UInt32.t) :
   Lemma (add_mod x (add_mod y z) == add_mod (add_mod x y) z)
   =
-  assert_norm (FStar.UInt.add_mod (v x) (FStar.UInt.add_mod (v y) (v z)) == FStar.UInt.add_mod (FStar.UInt.add_mod (v x) (v y)) (v z));
-  v_inj (add_mod x (add_mod y z)) (add_mod (add_mod x y) z)
+  let open Lib.IntTypes in
+  calc (==) {
```

Comment:
This can be solved by `canon_semiring` given an appropriate ring instance (see https://github.com/FStarLang/FStar/blob/master/examples/semiring/CanonCommSemiring.Test.fst#L57)
It may be worth if this kind of reasoning is common.